### PR TITLE
Add type hints to utilities.

### PR DESCRIPTION
**Pull request by jesnie**

### CODE REVIEW

File: gpflow/utilities/multipledispatch.py [hunk 1]
```diff
@@ -61,7 +70,7 @@ def get_first_occurrence(self, *types: Types) -> Callable[..., Any]:
         """
         n = len(types)
         for signature in self.ordering:
-            if len(signature) == n and all(map(issubclass, types, signature)):
+            if len(signature) == n and all(map(issubclass, types, signature)):  # type: ignore
```
Comment (uri-granta) [inline gpflow/utilities/multipledispatch.py:73]: (out of curiosity, what does this complain about?)

**Review by uri-granta — APPROVED** [review]
LGTM!

File: gpflow/utilities/multipledispatch.py (hunk 1 above)
Comment (jesnie) [reply gpflow/utilities/multipledispatch.py:75]: I forget. `issubclass` had some funny type expectations, and I gave up on making that work with `map` and `Types`.

File: gpflow/utilities/misc.py [hunk 2]
```diff
@@ -60,10 +60,10 @@ def is_variable(t: TensorData) -> bool:
 def training_loop(
     closure: Callable[[], tf.Tensor],
     optimizer: Optional[tf.optimizers.Optimizer] = None,
-    var_list: List[tf.Variable] = None,
-    maxiter=1e3,
-    compile=False,
-):
+    var_list: Optional[List[tf.Variable]] = None,
+    maxiter: int = int(1e3),
```
[the remark this answers is not in the document]
Comment (jesnie) [reply gpflow/utilities/misc.py]: > Any reason not to just write `1_000`?

You're right. I'll update it.

**Pull request merged**